File: julia/src/models/cdc.jl
"""CDC model — Change Data Capture via CDC_* SQL functions."""

struct CDCModel
    conn::LibPQ.Connection
    features::NucleusFeatures
end

"""CDC_READ(after_sequence, limit) → String (JSON change events)

Sent ONE argument, and the engine requires two: every call failed with
"CDC_READ requires (after_sequence, limit)". The function had never worked, and
nothing noticed because nothing executed it against a live engine.
"""
function cdc_read(m::CDCModel, after_sequence::Int64, limit::Int64=100)
    require_nucleus(m.features, "CDC")
    result = LibPQ.execute(m.conn, "SELECT CDC_READ(\$1, \$2)", [after_sequence, limit])
    val = first(result)[1]
    return _cdc_events(val)
end

"""Decode the engine's event array into a Vector of Dicts.

Returned the raw JSON string, so a caller had to parse it and the cross-SDK case
asserting a list passed against a non-empty string — a non-empty string is
truthy. Go and Rust were fixed the same way on 2026-08-15.

An empty result is an empty vector, never an error: "no changes since that
sequence" is the common case. A malformed payload IS an error, because silently
reporting "no changes" for something unparseable is the bug class CDC exists to
detect.
"""
function _cdc_events(raw)
    (ismissing(raw) || raw === nothing || isempty(raw)) && return Dict{String, Any}[]
    return [Dict{String, Any}(String(k) => v for (k, v) in pairs(e)) for e in JSON3.read(raw)]
end

"""CDC_COUNT() → Int64"""
function cdc_count(m::CDCModel)::Int64
    require_nucleus(m.features, "CDC")
    result = LibPQ.execute(m.conn, "SELECT CDC_COUNT()")
    return _int(result)
end

"""CDC_TABLE_READ(table, offset) → String (JSON change events for table)"""
function cdc_table_read(m::CDCModel, table::String, offset::Int64)::String
    require_nucleus(m.features, "CDC")
    result = LibPQ.execute(m.conn, "SELECT CDC_TABLE_READ(\$1, \$2)", [table, offset])
    val = first(result)[1]
    return ismissing(val) ? "[]" : val
end

File: rust/crates/neutron-nucleusdb/src/models/cdc.rs
//! Change Data Capture model — CDC_READ, CDC_COUNT, CDC_TABLE_READ.

use serde::{Deserialize, Serialize};

use crate::error::NucleusError;
use crate::pool::NucleusPool;
use crate::row_ext::RowExt;

/// A single change-data-capture log entry as emitted by the engine.
///
/// `read` and `table_read` returned the engine's raw JSON string and there was
/// no event type at all, so every caller wrote its own deserialization — and
/// the cross-SDK conformance case asserting a list passed against a non-empty
/// string, because a non-empty string is truthy. Python and TypeScript both
/// returned parsed events; Go and Rust did not.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CdcEvent {
    /// Monotonic sequence number of the change.
    #[serde(default)]
    pub seq: i64,
    /// Table the change applies to.
    #[serde(default)]
    pub table: String,
    /// Kind of change: `INSERT`, `UPDATE` or `DELETE`.
    #[serde(default)]
    pub change: String,
    /// Timestamp of the change, in epoch milliseconds.
    #[serde(default)]
    pub ts: i64,
}

/// Decode the engine's event array.
///
/// An empty result is an empty vector, never an error: "no changes since that
/// sequence" is the common case, not a failure. A malformed payload IS an error
/// rather than an empty vector, because silently reporting "no changes" when
/// the engine said something unparseable is the shape of bug this model exists
/// to detect.
fn parse_events(raw: &str) -> Result<Vec<CdcEvent>, NucleusError> {
    if raw.is_empty() {
        return Ok(Vec::new());
    }
    serde_json::from_str(raw).map_err(|e| NucleusError::Serde(e.to_string()))
}

/// Handle for CDC (Change Data Capture) operations.
pub struct CdcModel {
    pool: NucleusPool,
}

impl CdcModel {
    pub(crate) fn new(pool: NucleusPool) -> Self {
        Self { pool }
    }

    /// Read up to `limit` CDC events after the given sequence number.
    pub async fn read(
        &self,
        after_sequence: i64,
        limit: i64,
    ) -> Result<Vec<CdcEvent>, NucleusError> {
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT CDC_READ($1, $2)", &[&after_sequence, &limit])
            .await
            .map_err(NucleusError::Query)?;
        parse_events(&row.get_ck::<String>(0)?)
    }

    /// Return the total number of CDC events.
    pub async fn count(&self) -> Result<i64, NucleusError> {
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT CDC_COUNT()", &[])
            .await
            .map_err(NucleusError::Query)?;
        row.get_ck::<i64>(0)
    }

    /// Read up to `limit` CDC events for a specific table after the given
    /// sequence number.
    pub async fn table_read(
        &self,
        table: &str,
        after_sequence: i64,
        limit: i64,
    ) -> Result<Vec<CdcEvent>, NucleusError> {
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one(
                "SELECT CDC_TABLE_READ($1, $2, $3)",
                &[&table, &after_sequence, &limit],
            )
            .await
            .map_err(NucleusError::Query)?;
        parse_events(&row.get_ck::<String>(0)?)
    }
}
